File: shared/FlixorKit/Sources/FlixorKit/Storage/ProfileStorage.swift
//
//  ProfileStorage.swift
//  FlixorKit
//
//  Profile-scoped storage key management
//  Enables data isolation between Plex Home profiles
//

import Foundation

// MARK: - ProfileStorage

/// Manages profile context for scoped storage keys
public class ProfileStorage {
    public static let shared = ProfileStorage()

    private var currentProfileId: String?
    private let lock = NSLock()

    private init() {}

    // MARK: - Profile Context

    /// Set the current profile ID for scoped storage
    public func setCurrentProfile(_ profileId: String?) {
        lock.lock()
        defer { lock.unlock() }
        currentProfileId = profileId
    }

    /// Get the current profile ID
    public func getCurrentProfile() -> String? {
        lock.lock()
        defer { lock.unlock() }
        return currentProfileId
    }

    /// Check if currently using a profile (not main account)
    public var isUsingProfile: Bool {
        getCurrentProfile() != nil
    }

    // MARK: - Key Scoping

    /// Get a profile-scoped key
    /// - Parameter baseKey: The base storage key (e.g., "trakt_tokens")
    /// - Returns: Profile-scoped key if profile is active, otherwise base key
    public func getProfileKey(_ baseKey: String) -> String {
        guard let profileId = getCurrentProfile() else {
            return baseKey
        }
        return "profile:\(profileId):\(baseKey)"
    }

    /// Get a profile-scoped secure storage key
    /// - Parameter baseKey: The base storage key
    /// - Returns: Profile-scoped secure key
    public func getSecureProfileKey(_ baseKey: String) -> String {
        guard let profileId = getCurrentProfile() else {
            return "secure:\(baseKey)"
        }
        return "secure:profile:\(profileId):\(baseKey)"
    }

    // MARK: - Profile Data Management

    /// Clear all data for a specific profile
    /// - Parameter profileId: The profile UUID to clear
    public func clearProfileData(profileId: String) {
        // This would be implemented by the storage classes
        // to clear all keys with the profile prefix
    }
}

// MARK: - Profile-Scoped Keys

public extension StorageKeys {
    /// Keys that should be profile-scoped
    static let profileScopedKeys: Set<String> = [
        traktTokens,
        watchlistProvider,
        preferredQuality,
        preferredPlayerBackend
    ]

    /// Keys that should remain global (not profile-scoped)
    static let globalKeys: Set<String> = [
        plexToken,
        plexUser,
        selectedServer,
        selectedConnection,
        clientId,
        activeProfileId
    ]

    /// Check if a key should be profile-scoped
    static func isProfileScoped(_ key: String) -> Bool {
        // By default, keys are profile-scoped unless explicitly global
        return !globalKeys.contains(key)
    }
}

File: apps/mobile/src/core/ProfileStorage.ts
/**
 * Profile-scoped storage utility
 *
 * This module manages profile context for storage operations.
 * When a profile is active, storage keys are automatically scoped to that profile.
 * Main account uses un-prefixed keys for backward compatibility.
 */

// Current active profile UUID (null = main account)
let currentProfileId: string | null = null;

/**
 * Set the current active profile
 * @param id - Profile UUID or null for main account
 */
export function setCurrentProfile(id: string | null): void {
  currentProfileId = id;
}

/**
 * Get the current active profile UUID
 * @returns Profile UUID or null if using main account
 */
export function getCurrentProfile(): string | null {
  return currentProfileId;
}

/**
 * Generate profile-scoped storage key
 * Main account uses original keys (no prefix) for backward compatibility.
 * Profiles use prefixed keys: `profile:{uuid}:{baseKey}`
 *
 * @param baseKey - Original storage key (e.g., 'flixor_app_settings')
 * @returns Scoped key (e.g., 'profile:abc123:flixor_app_settings')
 */
export function getProfileKey(baseKey: string): string {
  if (!currentProfileId) {
    return baseKey; // Main account uses original keys
  }
  return `profile:${currentProfileId}:${baseKey}`;
}

/**
 * Check if a key is profile-scoped
 * @param key - Storage key to check
 * @returns True if the key is profile-scoped
 */
export function isProfileScopedKey(key: string): boolean {
  return key.startsWith('profile:');
}

/**
 * Extract profile ID from a scoped key
 * @param key - Profile-scoped key
 * @returns Profile UUID or null if not a scoped key
 */
export function extractProfileId(key: string): string | null {
  if (!isProfileScopedKey(key)) {
    return null;
  }
  const parts = key.split(':');
  return parts.length >= 2 ? parts[1] : null;
}

/**
 * Storage keys that should be profile-scoped
 */
export const PROFILE_SCOPED_KEYS = {
  // App settings
  APP_SETTINGS: 'flixor_app_settings',

  // Secure storage keys
  TRAKT_TOKENS: 'trakt_tokens',

  // Note: plex_auth is handled specially - stores main token + profile info
} as const;

/**
 * Storage keys that are global (shared across profiles)
 */
export const GLOBAL_KEYS = {
  // Client identification
  CLIENT_ID: 'flixor_client_id',

  // Onboarding state
  ONBOARDING_COMPLETED: 'flixor_onboarding_completed',

  // Main Plex auth (contains main token + current profile info)
  PLEX_AUTH: 'plex_auth',

  // Active profile tracking
  ACTIVE_PROFILE: 'flixor_active_profile',

  // Migration version
  MIGRATION_VERSION: 'flixor_migration_version',
} as const;
